RFC: Embrace NET CLI templates

nils-a opened the pull request.

nils-a committed:
add rfc proposal to embrace NET CLI templates

gep13 commented on text/0000-embrace-NET-CLI-templates.md:
```diff
@@ -0,0 +1,99 @@
+
+### Drawbacks to embracing NET CLI templates
+
+Compared to the possibilities of yeoman templates, NET CLI templates are rather "crude": The yeoman generator can create all item-templates in one go, as well as use the newest sources from a github repository.
```
```suggestion
Compared to the possibilities of yeoman templates, .NET CLI templates are rather "crude": The yeoman generator can create all item-templates in one go, as well as use the newest sources from a GitHub repository.
```

gep13 commented on text/0000-embrace-NET-CLI-templates.md:
```diff
@@ -0,0 +1,99 @@
+
+Compared to the possibilities of yeoman templates, NET CLI templates are rather "crude": The yeoman generator can create all item-templates in one go, as well as use the newest sources from a github repository.
+
+This features we would (probably, not 100% sure) lose.
```
```suggestion
These features we would (probably, not 100% sure) lose.
```

gep13 commented on text/0000-embrace-NET-CLI-templates.md:
```diff
@@ -0,0 +1,99 @@
+
+This features we would (probably, not 100% sure) lose.
+
+### Drawbacks to using NET CLI templates as the single source of templates in VS/Rider
```
```suggestion
### Drawbacks to using .NET CLI templates as the single source of templates in VS/Rider
```

gep13 commented on text/0000-embrace-NET-CLI-templates.md:
```diff
@@ -0,0 +1,99 @@
+
+Firstly, .NET CLI templates are "the way" of the .NET world, so we should supply all required project and item templates as .NET CLI templates.
+
+Secondly, maintaining one common set of templates is more "cost-effective" that maintaining multiple.
```
```suggestion
Secondly, maintaining one common set of templates is more "cost-effective" than maintaining multiple.
```

gep13 commented on text/0000-embrace-NET-CLI-templates.md:
```diff
@@ -0,0 +1,99 @@
+[unresolved-questions]: #unresolved-questions
+
+* Not sure when the feature in VS will leave experimental status
+* Not sure if showing item templates in VS is not supported of simply "currently broken"
```
```suggestion
* Not sure if showing item templates in VS is not supported or simply "currently broken"
```

gep13 commented on text/0000-embrace-NET-CLI-templates.md:
```diff
@@ -0,0 +1,99 @@
+* Not sure when the feature in VS will leave experimental status
+* Not sure if showing item templates in VS is not supported of simply "currently broken"
+* Not sure if Rider will start supporting item templates
+* could NET CLI templates provide templates to be downloaded, as cake-yeoman currently does.
```
```suggestion
* Could .NET CLI templates provide templates to be downloaded, as cake-yeoman currently does.
```

gep13 commented:
@nils-a what about VSCode? Does a .NET CLI template "appear" in here as well?  We currently have a command in the VSCode extension to setup/add Cake related items to a workspace. Should we consider replacing that with this as well?

nils-a committed:
fixed spelling

nils-a committed:
added Cake for VS Code

in the motivation and design sections.

nils-a committed:
more spelling mistakes.

nils-a commented:
@gep13 :

> what about VSCode? Does a .NET CLI template "appear" in here as well?

No, VSCode does not have the notion of project or item templates. 

> Should we consider replacing that with this as well?

Yes, definitively. As you said, we're currently using commands to simulate item templates. We could simply modify those commands internally, so that .NET CLI templates are used instead of the current flow. (The current flow being: There is a template for `build.cake`, the rest is downloaded on-the-fly.) We should probably consider installing/updating the templates package, as needed - the same way we do for the Cake runner.

I added *Cake for VS Code* to the motivations section, as this is one more template we're maintaining currently. I also added changing the inner workings of the commands in *Cake for VS Code* to the design section. (Not the auto-installation, though. I feel that's purely for consistency with the other workflows in *Cake for VS Code*.)

nils-a committed:
added rider specific links

esp. https://youtrack.jetbrains.com/issue/RIDER-24197 - the
suggestion to implement .NET CLI item templates in rider.

nils-a commented:
I did some tests.

CLI currently looks like this:
```text
❯ dotnet new -u
...
  Cake.Templates
    Details:
      NuGetPackageId: Cake.Templates
      Version: 0.1.0
      Author: Patrik Svensson, Mattias Karlsson, Gary Ewan Park, Alistair Chapman, Martin Björkström, Dave Glick, Pascal Berger, Jérémie Desautels, Enrico Campidoglio, C. Augusto Proiete, Nils Andresen, and contributors
    Templates:
      Cake Addin testproject (cakeaddintest) C#
      Cake Addin (cakeaddin) C#
      Cake Bootstrapper (cakebootstrap) PS
      Cake Bootstrapper (cakebootstrap) sh
      Cake Build file (cakebuild)
      Cake Frosting Build Project (cakefrosting) C#
      Cake Module (cakemodule) C#
    Uninstall Command:
      dotnet new -u Cake.Templates

❯ dotnet new cake --list
These templates matched your input: 'cake'.

Templates                    Short Name     Language  Tags
---------------------------  -------------  --------  ---------------------------------
Cake Frosting Build Project  cakefrosting   [C#]      Cake/Frosting/Windows/Linux/macOs
Cake Addin testproject       cakeaddintest  [C#]      Cake/Windows/Linux/macOs
Cake Addin                   cakeaddin      [C#]      Cake/Windows/Linux/macOs
Cake Bootstrapper            cakebootstrap  PS,sh     Cake/Windows/Linux/macOs
Cake Module                  cakemodule     [C#]      Cake/Windows/Linux/macOs
Cake Build file              cakebuild                Cake/Windows/Linux/macOs
```

Some of these have options, e.g. the two "test" projects from cake-vs are integrated in `cakeaddintest` via an option `--empty`.

In VS that looks like:
![image](https://user-images.githubusercontent.com/349188/117881458-8f665e80-b2a9-11eb-86bf-5839621f4b91.png)

with all options available, the same as in CLI:
![image](https://user-images.githubusercontent.com/349188/117881845-013ea800-b2aa-11eb-8704-0b3aedc306f8.png)

Rider sadly is not so perfect:
While it is possible to add icons to CLI templates and also rename the template. They will currently be listed under the category `Other` and will **not** have options available

![image](https://user-images.githubusercontent.com/349188/117882722-e9b3ef00-b2aa-11eb-8cfa-b2df2e7d4646.png)

I have not looked in VS Code, but since that's all custom code anyway I expect no surprises there.

nils-a commented:
Update: Adding a .NET component to the rider plugin (i.e. a ReSharper part) it is possible to get at least some parameters working. (And in fact this seems to be the "way" to provide solution templates in Rider.)

![image](https://user-images.githubusercontent.com/349188/118412404-016be880-b69a-11eb-9234-729b4a596d23.png)

One downside is, that the templates (i.e. the `content` of the nupkg) needs to be included in the plugin.

pascalberger commented:
I recently played with the idea of providing a template for Cake.Issue.Recipe. For Frosting, this would bascially be the Frosting template with Cake.Frosting.Issues.Recipe added and configured. 

Do we also aim to integrate these community extensions into a single template (I don't think .NET CLI templates are advanced enough to be capable to handle these through parameters or inheritance)?

nils-a commented:
@pascalberger up to now, I've only thought about harmonizing the templates that we currently ship to keep the need for maintaining multiple template forms and versions to a minimum.

gep13 commented:
The RFC process is no longer being followed for Cake. As such, I am going to go this pull request.
